File: backend/src/protocol.ts
// 手机端 ↔ 后端 的 WebSocket 消息协议（JSON，每条消息都有 type 字段）
// 与 Android 端 data/Protocol.kt 保持一一对应。

export type PermissionMode =
  | "default"
  | "acceptEdits"
  | "plan"
  | "dontAsk"
  | "bypassPermissions";

// ────────────── 客户端 → 服务端 ──────────────
export type ClientMessage =
  | { type: "auth"; token: string }
  | { type: "send_message"; prompt: string; sessionId?: string; permissionMode?: PermissionMode }
  | { type: "interrupt" }
  | { type: "permission_response"; requestId: string; decision: "allow" | "deny"; updatedInput?: unknown }
  | { type: "list_sessions"; dir?: string }
  | { type: "load_session"; sessionId: string }
  | { type: "set_permission_mode"; mode: PermissionMode }
  | { type: "set_cwd"; cwd: string };

// ────────────── 服务端 → 客户端 ──────────────
export type ServerMessage =
  | { type: "hello"; ok: true; serverVersion: string }
  | { type: "error"; message: string; code?: string }
  | { type: "system"; sessionId: string; tools: string[]; model: string; cwd: string; availableCwds?: string[] }
  | { type: "assistant"; messageId: string; content: unknown[] }
  | { type: "assistant_partial"; textDelta: string }
  | { type: "thinking_progress"; estimatedTokens: number; delta: number }
  | { type: "tool_use"; toolUseId: string; toolName: string; input: unknown }
  | { type: "tool_result"; toolUseId: string; content: unknown; isError: boolean }
  | { type: "permission_request"; requestId: string; toolName: string; input: unknown; summary: string }
  | { type: "result"; subtype: string; costUsd?: number; usage?: unknown; terminalReason?: string; isError: boolean }
  | { type: "status"; message: string }
  | { type: "session_list"; sessions: unknown[] }
  | { type: "session_messages"; sessionId: string; messages: unknown[] };

File: android/app/src/main/java/com/example/clauderemote/data/Protocol.kt
package com.example.clauderemote.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
enum class PermissionMode {
    default, acceptEdits, plan, dontAsk, bypassPermissions
}

/** assistant 消息 content 数组里的一个块（text / tool_use / thinking） */
@Serializable
data class ContentBlock(
    val type: String,
    val text: String? = null,
    val id: String? = null,
    val name: String? = null,
    val input: JsonElement? = null,
    val thinking: String? = null,
)

// ────────────── 客户端 → 服务端 ──────────────
@Serializable
sealed class ClientMessage {
    @Serializable @SerialName("auth")
    data class Auth(val token: String) : ClientMessage()

    @Serializable @SerialName("send_message")
    data class SendMessage(
        val prompt: String,
        @SerialName("sessionId") val sessionId: String? = null,
        @SerialName("permissionMode") val permissionMode: PermissionMode? = null,
    ) : ClientMessage()

    @Serializable @SerialName("interrupt")
    object Interrupt : ClientMessage()

    @Serializable @SerialName("permission_response")
    data class PermissionResponse(
        val requestId: String,
        val decision: String,
        val updatedInput: JsonElement? = null,
    ) : ClientMessage()

    @Serializable @SerialName("set_permission_mode")
    data class SetPermissionMode(val mode: PermissionMode) : ClientMessage()

    @Serializable @SerialName("set_cwd")
    data class SetCwd(val cwd: String) : ClientMessage()

    @Serializable @SerialName("list_sessions")
    data class ListSessions(val dir: String? = null) : ClientMessage()

    @Serializable @SerialName("load_session")
    data class LoadSession(val sessionId: String) : ClientMessage()
}

// ────────────── 服务端 → 客户端 ──────────────
@Serializable
sealed class ServerMessage {
    @Serializable @SerialName("hello")
    data class Hello(
        val ok: Boolean,
        val serverVersion: String,
        val cwd: String? = null,
        val availableCwds: List<String> = emptyList(),
    ) : ServerMessage()

    @Serializable @SerialName("error")
    data class Error(val message: String, val code: String? = null) : ServerMessage()

    @Serializable @SerialName("system")
    data class SystemInit(
        val sessionId: String,
        val tools: List<String>,
        val model: String,
        val cwd: String,
        val availableCwds: List<String> = emptyList(),
    ) : ServerMessage()

    @Serializable @SerialName("assistant")
    data class Assistant(val messageId: String, val content: List<ContentBlock>) : ServerMessage()

    @Serializable @SerialName("assistant_partial")
    data class AssistantPartial(val textDelta: String) : ServerMessage()

    @Serializable @SerialName("tool_use")
    data class ToolUse(val toolUseId: String, val toolName: String, val input: JsonElement) : ServerMessage()

    @Serializable @SerialName("tool_result")
    data class ToolResult(val toolUseId: String, val content: JsonElement, val isError: Boolean) : ServerMessage()

    @Serializable @SerialName("permission_request")
    data class PermissionRequest(
        val requestId: String,
        val toolName: String,
        val input: JsonElement,
        val summary: String,
    ) : ServerMessage()

    @Serializable @SerialName("thinking_progress")
    data class ThinkingProgress(val estimatedTokens: Int, val delta: Int) : ServerMessage()

    @Serializable @SerialName("result")
    data class Result(
        val subtype: String,
        val costUsd: Double? = null,
        val usage: JsonElement? = null,
        val terminalReason: String? = null,
        val isError: Boolean = false,
    ) : ServerMessage()

    @Serializable @SerialName("status")
    data class Status(val message: String) : ServerMessage()

    @Serializable @SerialName("session_list")
    data class SessionList(val sessions: List<JsonElement>) : ServerMessage()

    @Serializable @SerialName("session_messages")
    data class SessionMessages(val sessionId: String, val messages: List<JsonElement>) : ServerMessage()
}
